chroma_core/plugins/block_devices.py

```diff
@@ -12,23 +12,51 @@
 log.setLevel(DEBUG)
 
 
-def _fetch_aggregator():
+def _fetch_aggregator(timeout):
     import requests
 
-    resp = requests.get(settings.DEVICE_AGGREGATOR_URL)
-    payload = resp.text
-
-    return json.loads(payload)
-
-
-def get_devices(fqdn):
-    try:
-        _data = _fetch_aggregator()
-        return _data[fqdn]
-    except Exception as e:
-        log.error(
-            "iml-device-aggregator is not providing expected data, ensure "
-            "iml-device-scanner package is installed and relevant "
-            "services are running on storage servers (%s)" % e
-        )
-        return {}
+    summary = 0.0
+    timeout = timeout if timeout >= 0 else 0
+    while summary <= timeout:
+        resp = requests.get(settings.DEVICE_AGGREGATOR_URL)
+        summary += resp.elapsed.seconds
+        summary += resp.elapsed.microseconds / 10.0 ** 6
+
+        try:
+            return resp.json()
+        except ValueError as e:
+            # This error might be casused by device-aggregator not being ready
+            # yet or communication breakdown inside IML server side
+            log.error(
+                "iml-device-aggregator is not providing expected data, ensure "
+                "this is not caused by race condition (%s)" % e
+            )
+            # So it is better to wait for service startup
+            if summary < timeout:
+                time.sleep(0.1)
+                summary += 0.1
+    return {}, summary
+
+
+def get_devices(fqdn, timeout=0):
+    summary = 0.0
+    timeout = timeout if timeout >= 0 else 0
+
+    while summary <= timeout:
+        try:
+            _data, elapsed = _fetch_aggregator(timeout - summary)
+            summary += elapsed
+            return _data[fqdn]
+        except (KeyError, TypeError) as e:
+            # This error might be caused by fact that device-aggregator
+            # hasn't yet received device information update on startup
+            log.error(
+                "iml-device-aggregator is not providing expected data, ensure "
+                "iml-device-scanner package is installed and relevant "
+                "services are running on storage servers (%s)" % e
```

Comment:
I know you didn't write this error message, but I think the wording here does not reflect what the probable causes are.

Something like:

`iml-device-aggregator is not providing expected data for fqdn {}. Waited: {}, Trying again for: {}, Error: {}.format(fqdn, summary, timeout - summary, e)`